Modules/posixmodule.c

```diff
@@ -10064,6 +10069,252 @@ os_times_impl(PyObject *module)
+        The initial expiration time, in seconds.
+
+        * If *flags* has TFD_TIMER_ABSTIME bit, *flags* must be in absolute time.
+        * If *flags* doesn't have TFD_TIMER_ABSTIME bit, *flags* must be in relative time.
+        * If *flags* has TFD_TIMER_ABSTIME bit and TFD_TIMER_CANCEL_ON_SET bit and *flags* and time.CLOCK_REALTIME
+          and system clock is changed discontinuously, reading a file descriptor is aborted with ECANCELED.
+    interval: double = 0.0
+        The timer's interval, in seconds.
+
+        * If 'interval' is zero, timer will be run once.
+        * If 'interval' is non-zero, timer will be run periodically with the interval.
+
+Alter a timer file descriptor's internal timer in seconds.
+[clinic start generated code]*/
+
+static PyObject *
+os_timerfd_settime_impl(PyObject *module, int fd, int flags, double initial,
+                        double interval)
+/*[clinic end generated code: output=0dda31115317adb9 input=436a433c2c9664f0]*/
+{
+    struct itimerspec new_value;
+    struct itimerspec old_value;
+    int result;
+    new_value.it_value.tv_sec = (time_t)initial;
+    new_value.it_value.tv_nsec = EXTRACT_NSEC(initial);
+    new_value.it_interval.tv_sec = (time_t)interval;
+    new_value.it_interval.tv_nsec = EXTRACT_NSEC(interval);
+    Py_BEGIN_ALLOW_THREADS
+    result = timerfd_settime(fd, flags, &new_value, &old_value);
+    Py_END_ALLOW_THREADS
+    if (result == -1) {
+        return PyErr_SetFromErrno(PyExc_OSError);
+    }
+    return build_itimerspec(&old_value);
+}
+
+
+/*[clinic input]
+os.timerfd_settime_ns
+
+    fd: fildes
+        A timer file descriptor.
+    /
+    *
+    flags: int = 0
+        0 or a bit mask of TFD_TIMER_ABSTIME or TFD_TIMER_CANCEL_ON_SET.
+    initial: long_long = 0
+        initial expiration timing in seconds.
+
+        * If *flags* has TFD_TIMER_ABSTIME bit, *flags* must be in absolute time.
+        * If *flags* doesn't have TFD_TIMER_ABSTIME bit, *flags* must be in relative time.
+        * If *flags* has TFD_TIMER_ABSTIME bit and TFD_TIMER_CANCEL_ON_SET bit and *flags* and time.CLOCK_REALTIME
+          and system clock is changed discontinuously, reading a file descriptor is aborted with ECANCELED.
+    interval: long_long = 0
+        interval for the timer in seconds.
+
+        * If 'interval' is zero, timer will be run once.
+        * If 'interval' is non-zero, timer will be run periodically with the interval.
+
+Alter a timer file descriptor's internal timer in nanoseconds.
+[clinic start generated code]*/
+
+static PyObject *
+os_timerfd_settime_ns_impl(PyObject *module, int fd, int flags,
+                           long long initial, long long interval)
+/*[clinic end generated code: output=6273ec7d7b4cc0b3 input=81eb655e9802cb05]*/
+{
+    struct itimerspec new_value;
+    struct itimerspec old_value;
+    int result;
+    new_value.it_value.tv_sec = initial / ONE_SECOND_IN_NS;
+    new_value.it_value.tv_nsec = initial % ONE_SECOND_IN_NS;
```

Comment:
Perhaps `_PyTime_FromNanoseconds()` or `_PyTime_FromNanosecondsObject()` followed by `_PyTime_AsTimespec()`/`_PyTime_AsTimespec_clamp()` could be used here to avoid integer overflow.

cc @vstinner

Reply:
Yeah, please use the _PyTime_t API which takes care of rounding, overflow errors, etc.